src/statsmodels/contrasts.jl

```diff
@@ -0,0 +1,233 @@
+## Specify contrasts for coding categorical data in model matrix. Contrasts types
+## are a subtype of AbstractContrasts. ContrastsMatrix types hold a contrast
+## matrix, levels, and term names and provide the interface for creating model
+## matrix columns and coefficient names.
+##
+## Contrasts types themselves can be instantiated to provide containers for
+## contrast settings (currently, just the base level).
+##
+## ModelFrame will hold a Dict{Symbol, ContrastsMatrix} that maps column
+## names to contrasts.
+##
+## ModelMatrix will check this dict when evaluating terms, falling back to a
+## default for any categorical data without a specified contrast.
+##
+## TODO: implement contrast types in Formula/Terms
+##
+## Dave Kleinschmidt 2015-2016
+
+
+"""
+Interface to describe contrast coding schemes for categorical variables.
+
+Concrete subtypes of `AbstractContrasts` describe a particular way of converting a
+categorical column in a `DataFrame` into numeric columns in a `ModelMatrix`. Each
+instantiation optionally includes the levels to generate columns for and the base
+level. If not specified these will be taken from the data when a `ContrastsMatrix` is
+generated (during `ModelFrame` construction).
+
+## Constructors
+
+For `C <: AbstractContrast`:
+
+```julia
+C()                             # levels are inferred later 
+C(levels = ::Vector{Any})       # levels checked against data later
+C(base = ::Any)                 # specify base level
+C(levels = ::Vector{Any}, base = ::Any)
+```
+
+If specified, levels will be checked against data when generating a
+`ContrastsMatrix`. Any mismatch will result in an error, because missing data
+levels would lead to empty columns in the model matrix, and missing contrast
+levels would lead to empty or undefined rows.
+
+You can also specify the base level of the contrasts. The
+actual interpretation of this depends on the particular contrast type, but in
+general it can be thought of as a "reference" level.  It defaults to the first
+level.
+
+Both `levels` and `base` will be coerced to the type of the data when
+constructing a `ContrastsMatrix`.
+
+## Concrete types
+
+* `TreatmentContrasts`
+* `SumContrasts`
+* `HelmertContrasts`
+
+To implement your own concrete types, implement a constructor and a method for
+constructing the actual contrasts matrix that maps from levels to `ModelMatrix`
+column values:
+
+```julia
+type MyContrasts <: AbstractContrasts
+    ...
+end
+
+contrasts_matrix(C::MyContrasts, baseind, n) = ...
+```
+
+"""
+abstract AbstractContrasts
+
+## Contrasts + Data = ContrastsMatrix
+type ContrastsMatrix{T}
+    matrix::Matrix{Float64}
+    termnames::Vector{T}
+    levels::Vector{T}
+    contrasts::AbstractContrasts
```

Comment:
I worry that by doing this you'd lose information about, say, whether levels were specified on the `AbstractContrast` itself. But that might not be a problem, given that the `ContrastMatrix` type is supposed to contain all the information necessary to actually construct `ModelMatrix` columns (and it's the thing that's stored in the `ModelFrame`).